mantidproject/mantid
# Construct transform on init of sliceinfo and remove circular dependence in sliceviewer

**Description of work.**

The transform uses the angle between the display axes. Before the transform was constructed in the model which required the viewing axes/display indices to be passed in the sliceinfo object. In addition the sliceinfo (which is used to get the transform for the matplotlib stuff and in the peaksviewer) was initialised with a transform stored in the data_view (which wasn't being kept updated with the display axes as it was only updated from the sliceinfo when nonorthog/orthog axes were created when the non-orthog view button is clicked).

Here we construct transform with angle between axes when the sliceinfo is init (using angles calculated in model using OrientedLattice on init). The transform in the data_view is only updated from the sliceinfo when non-orthog axes are created (so no longer any circular dependence - the transform is not passed back from data_view to sliceinfo).

The enabling of the non-orthogonal view button on opening is now done by querying the sliceinfo object - as one needs to know the display indices.

Also for MDHisto workspaces it now gets the projection matrix from the basis vectors on the workspace rather than search for the W_MATRIX log (which doesn't exist causing it to assume an identity matrix)



**To test:**

(1) Repeat tests in PR #32392 (check that non-ortho button is enabled when expected for various 2D,3D and 4D workspaces)

- To check the transform of a 2D MD workspace (technically it's still 3D but you can't edit the slice in the third dimentsion), open e.g. `ws_nonortho_svrebinned` in sliceviewer from the above

(2) Follow instructions in #32436 and check that 
- slices with energy have the non-ortho view disabled
- when viewing HK plane that non-ortho view is enabled and the transform is OK
- that switching from e.g. (x,y)=(H,K) to (x,y)=(H,Energy) that non-ortho view is disabled

(3) Check transform is correct when UB replaced or deleted e.g.
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')

expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
# i) open ws in sliceviewer
# ii) click non-orthog transform view
# iii) Change UB to orthogonal (note the plot won't update automatically)
SetUB(ws, 1,1,2,90,90,90)
# iv) Toggle non-ortho view - it should be correct now
```
(4) Check that when the frame is changed such that the support for non-ortho transform is different then sliceviewer closes with error
`The property supports_nonorthogonal_axes is different on the new workspace.`
To do this e.g. 
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
FakeMDEventData(ws, EllipsoidParams='1e+04,0,1,0,1,0,0,0,1,0,0,0,1,0.2,0.6,0.3,1', RandomSeed='3873875')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)

# open ws in sliceviewer here

ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
               Names='Q1,Q2,Q3', Units='Ang-1,Ang-1,Ang-1',
               Frames='QLab,QLab,QLab',
               SplitInto='2', SplitThreshold='50')    
```
(5) Check that peaks can be overlaid (open sliveviewer > three blobs button > select `peaks`)
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-2,2',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
FakeMDEventData(ws, EllipsoidParams='1e+04,0,1,1,1,0,0,0,1,0,0,0,1,0.02,0.06,0.03,1', RandomSeed='3873875')
# add peak
CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
CopySample(InputWorkspace='ws', OutputWorkspace='peaks', CopyName=False, CopyMaterial=False, CopyEnvironment=False, CopyShape=False)
p = mtd['peaks']
pk = p.createPeakHKL([0,1,1]) # axes aligned with viewing axes
p.addPeak(pk)

IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
    BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int_centroid_unfixed', 
    Ellipsoid=True, UseCentroid=False, FixMajorAxisLength = True, UseOnePercentBackgroundCorrection=False)
```

Try toggling non-ortho view on and off - note that the position of the peak and the ellipsoid shape of the integrated peak should agree with the data.

(6) Check that the add/delete peak functionality works (with non-ortho view on and off)
(7) Test with workspace with other frames - e.g. QLab using this workspace
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
               Names='Q1,Q2,Q3', Units='Ang-1,Ang-1,Ang-1',
               Frames='QLab,QLab,QLab',
               SplitInto='2', SplitThreshold='50')   
```
(8) Check that non-axis aligned slices can be viewed in sliceviewer correctly
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='10')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
BinMD(InputWorkspace=ws, AxisAligned=False,
    BasisVector0='[00L],r.l.u.,0,0,1', 
    BasisVector1='[HH0],r.l.u.,1,1,0',
    BasisVector2='[-HH0],r.l.u.,-1,1,0',
    OutputExtents='-4,4,-4,4,-0.25,0.25', 
    OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
```
Check `ws_slice` has non-ortho view enabled bu when it is clicked that the grid is orthogonal (in this case 110 is orthogonal to 001).

Fixes #32436 and #32469  (and permanent fix for #32314 - for which a temp. fix was done in PR #32392)






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
